### ISSUE TITLE
match_cert: Wrong certificate returned if multiple certs have same label but different ID

### ISSUE THREAD

**Issue by minfrin** [OpenSC/libp11#435]
Using the new debug logging in https://github.com/OpenSC/libp11/pull/433 I have found an edge case where the wrong certificate is returned when matching certs.

In the example below we look for a cert with the ID b0905a361ae02fda131481fce1de9d869e308fae, but we return a cert with the ID 8d8c5ec454ad8ae177e99bf99b05e1b8018d61e1. OpenSSL then correctly complains the cert and key do not match.

```
Looking in slots for certificate without login: id=b0905a361ae02fda131481fce1de9d869e308fae label=
- [17] Linux                      login                                 (Local Tokens)
- [18] SoftHSM slot ID 0x6be9cd2  login                                 (SoftHSM Token)
- [19] SoftHSM slot ID 0x1        uninitialized, login                  (no label)
Found slot:  Linux
Found token: Local Tokens
Found 2 certificates:
   1    id=b0905a361ae02fda131481fce1de9d869e308fae label=
   2    id=8d8c5ec454ad8ae177e99bf99b05e1b8018d61e1 label=
Returning last matching certificate: id=8d8c5ec454ad8ae177e99bf99b05e1b8018d61e1 label=
```

The cause is here:

https://github.com/OpenSC/libp11/blob/master/src/eng_back.c#L626

```
			if (obj_label && k->label && strcmp(k->label, obj_label) == 0)
				selected_cert = k;
			if (obj_id_len != 0 && k->id_len == obj_id_len &&
					memcmp(k->id, obj_id, obj_id_len) == 0)
				selected_cert = k;
```

Most specifically, the above code says if the labels match OR the ids match, we have a match.

In this example, both certs have an empty label (for whatever reason), so we consider this a match when we should not.

Happy to fix this, but it with conflict with https://github.com/OpenSC/libp11/pull/433.

Must I fix this in https://github.com/OpenSC/libp11/pull/433, or should I wait until this is applied and fix it after?

**Comment by minfrin** [comment]
The fix looks like this below:

Github can't figure out how to handle patch layered on top of a patch, so https://github.com/OpenSC/libp11/pull/433 needs to go first.

```
diff --git a/src/eng_back.c b/src/eng_back.c
index 0372afe..c06eba2 100644
--- a/src/eng_back.c
+++ b/src/eng_back.c
@@ -636,11 +636,22 @@ static void *match_cert(ENGINE_CTX *ctx, PKCS11_TOKEN *tok,
 			dump_hex(ctx, 1, k->id, k->id_len);
 			ctx_log(ctx, 1, " label=%s\n", k->label ? k->label : "(null)");
 
-			if (obj_label && k->label && strcmp(k->label, obj_label) == 0)
-				selected_cert = k;
-			if (obj_id_len != 0 && k->id_len == obj_id_len &&
-					memcmp(k->id, obj_id, obj_id_len) == 0)
-				selected_cert = k;
+			if (obj_label && obj_id_len != 0) {
+				if (k->label && strcmp(k->label, obj_label) == 0 &&
+						k->id_len == obj_id_len &&
+						memcmp(k->id, obj_id, obj_id_len) == 0) {
+					selected_cert = k;
+				}
+			} else if (obj_label && !obj_id_len) {
+				if (k->label && strcmp(k->label, obj_label) == 0) {
+					selected_cert = k;
+				}
+			} else if (obj_id_len && !obj_label) {
+				if (k->id_len == obj_id_len &&
+						memcmp(k->id, obj_id, obj_id_len) == 0) {
+					selected_cert = k;
+				}
+			}
 		}
 	} else {
 		which = "first (with id present)";
@@ -729,11 +740,22 @@ static void *match_key(ENGINE_CTX *ctx, const char *key_type,
 			dump_hex(ctx, 1, k->id, k->id_len);
 			ctx_log(ctx, 1, " label=%s\n", k->label ? k->label : "(null)");
 
-			if (obj_label && k->label && strcmp(k->label, obj_label) == 0)
-				selected_key = k;
-			if (obj_id_len != 0 && k->id_len == obj_id_len
-					&& memcmp(k->id, obj_id, obj_id_len) == 0)
-				selected_key = k;
+			if (obj_label && obj_id_len != 0) {
+				if (k->label && strcmp(k->label, obj_label) == 0 &&
+						k->id_len == obj_id_len &&
+						memcmp(k->id, obj_id, obj_id_len) == 0) {
+					selected_key = k;
+				}
+			} else if (obj_label && !obj_id_len) {
+				if (k->label && strcmp(k->label, obj_label) == 0) {
+					selected_key = k;
+				}
+			} else if (obj_id_len && !obj_label) {
+				if (k->id_len == obj_id_len &&
+						memcmp(k->id, obj_id, obj_id_len) == 0) {
+					selected_key = k;
+				}
+			}
 		}
 	} else {
 		which = "first";
```

**Comment by minfrin** [comment]
Github can't handle layered patches.

Added fix to https://github.com/OpenSC/libp11/pull/433

**Comment by mtrojnar** [comment]
Merged with #433
